### PR TITLE
duplicate shift

### PR DESCRIPTION
**Pull request by Sai9797**
Fixes #issue_number

**What was changed?**

*Here, describe what part of the application you changed (e.g. login page, database, etc.). If possible, mention what specific components were added, removed, or modified.*

**Why was it changed?**

*Here, describe the issue that you are fixing with this code, and why your code fixes it.*

**How was it changed?**

*Here, get into detail about what files you modified, and talk about the most important lines in regards to fixing the issue.*

**Screenshots that show the changes (if applicable):**

### CODE REVIEW

**Comment by Sai9797** [comment]
Fixes #21

What was changed?

Modify adding the shift to check if there are any overlapping in the shift timings.

Why was it changed?

The server should not allow the same user to sign up for the same shift more than one time or to sign up for a work-shift that overlaps with another shift that the same user is scheduled to work.

How was it changed?

Changed workshift_add_use_case and workshift_add_multiple_use_case methods along with POST api call.

Screenshots that show the changes (if applicable):
image

**Review by kate-holdener — CHANGES_REQUESTED** [review]
I selected two identical shifts through the client side app and submitted my request to sign up. The server showed the following errors:
```
127.0.0.1 - - [04/Dec/2023 10:36:59] "POST /shifts HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 1478, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 1458, in wsgi_app
    response = self.handle_exception(e)
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 1455, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 869, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask_cors/extension.py", line 176, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 867, in full_dispatch_request
    rv = self.dispatch_request()
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 852, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask_cors/decorator.py", line 130, in wrapped_function
    resp = make_response(f(*args, **kwargs))
  File "/Users/holdenerea/repos/shelter_volunteers/server/application/rest/work_shift.py", line 145, in work_shifts
    add_responses = workshift_add_multiple_use_case(repo,
  File "/Users/holdenerea/repos/shelter_volunteers/server/use_cases/add_workshifts.py", line 26, in workshift_add_multiple_use_case
    add_response = workshift_add_use_case(repo, work_shift, existing_shifts)
  File "/Users/holdenerea/repos/shelter_volunteers/server/use_cases/add_workshifts.py", line 12, in workshift_add_use_case
    if shift_already_exists(new_shift, existing_shifts):
  File "/Users/holdenerea/repos/shelter_volunteers/server/use_cases/add_workshifts.py", line 42, in shift_already_exists
    existing_start = convert_timestamp_to_datetime(shift.start_time)
AttributeError: 'dict' object has no attribute 'start_time'
```
Then, I submitted those shifts again and it seemed to work. The server response is 409, and one of the shifts was successfully added. 

If there is a conflict in selected shifts, I think we shouldn't add any shifts, but respond to the client with details about the conflict. @AbhilashKotha @loganwyas what are your thoughts on this?

File: server/tests/use_cases/test_add_workshifts.py
Comment (kate-holdener) [inline server/tests/use_cases/test_add_workshifts.py:1]: You should add test cases here to test for attempts to add conflicting shifts.

**Comment by kate-holdener** [comment]
> I selected two identical shifts through the client side app and submitted my request to sign up. The server showed the following errors:
> 
> ```
> 127.0.0.1 - - [04/Dec/2023 10:36:59] "POST /shifts HTTP/1.1" 500 -
> Traceback (most recent call last):
>   File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 1478, in __call__
>     return self.wsgi_app(environ, start_response)
>   File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 1458, in wsgi_app
>     response = self.handle_exception(e)
>   File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask_cors/extension.py", line 176, in wrapped_function
>     return cors_after_request(app.make_response(f(*args, **kwargs)))
>   File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 1455, in wsgi_app
>     response = self.full_dispatch_request()
>   File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 869, in full_dispatch_request
>     rv = self.handle_user_exception(e)
>   File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask_cors/extension.py", line 176, in wrapped_function
>     return cors_after_request(app.make_response(f(*args, **kwargs)))
>   File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 867, in full_dispatch_request
>     rv = self.dispatch_request()
>   File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask/app.py", line 852, in dispatch_request
>     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
>   File "/Users/holdenerea/repos/shelter_volunteers/server/venv/lib/python3.9/site-packages/flask_cors/decorator.py", line 130, in wrapped_function
>     resp = make_response(f(*args, **kwargs))
>   File "/Users/holdenerea/repos/shelter_volunteers/server/application/rest/work_shift.py", line 145, in work_shifts
>     add_responses = workshift_add_multiple_use_case(repo,
>   File "/Users/holdenerea/repos/shelter_volunteers/server/use_cases/add_workshifts.py", line 26, in workshift_add_multiple_use_case
>     add_response = workshift_add_use_case(repo, work_shift, existing_shifts)
>   File "/Users/holdenerea/repos/shelter_volunteers/server/use_cases/add_workshifts.py", line 12, in workshift_add_use_case
>     if shift_already_exists(new_shift, existing_shifts):
>   File "/Users/holdenerea/repos/shelter_volunteers/server/use_cases/add_workshifts.py", line 42, in shift_already_exists
>     existing_start = convert_timestamp_to_datetime(shift.start_time)
> AttributeError: 'dict' object has no attribute 'start_time'
> ```
> 
> Then, I submitted those shifts again and it seemed to work. The server response is 409, and one of the shifts was successfully added.
> 
> If there is a conflict in selected shifts, I think we shouldn't add any shifts, but respond to the client with details about the conflict. @AbhilashKotha @loganwyas what are your thoughts on this?

@Sai9797 This issue still needs to be addressed.

Also, as we discussed after your presentation, the response to the request to add shifts should contain a list of shift ids (the ids that were sent with the request) and status (added or not), and an error message (if shift was not added), detailing why the shift was not added. For example, an example response payload could be:
```
{
   {
      "id": "abc",
      "success": true,
   },
   {
      "id": "xyz",
      "success": false,
      "error": "You are signed up for another shift at this time"
   },
}
```

**Comment by loganwyas** [comment]
I do think that not signing up any shifts if a conflict occurs would be best. Also, I don't know if this is something that has already been addressed or would be better for a separate issue, but do we want to have something (either on the client side or the server side) that prevents conflicting shifts within the same request (where none of them have been added to the server yet)?

**Comment by AbhilashKotha** [comment]
We can directly prevent adding it to the selection instead of taking them to the next step and showing an error.

**Comment by kate-holdener** [comment]
> We can directly prevent adding it to the selection instead of taking them to the next step and showing an error.

I agree, we can and should do that. However, server should prevent addition of any conflicting shifts, regardless of how the conflict is created. Whether it's two conflicting shifts in the request (even though we'll prevent it from the client side) or if it's a request to sign up for a shift, that conflicts with an existing shift.

**Comment by AbhilashKotha** [comment]
I think we should stick to the server-side validation with PR and create a new issue for client side

**Comment by kate-holdener** [comment]
> I think we should stick to the server-side validation with PR and create a new issue for client side

I agree with you here.

File: server/use_cases/add_workshifts.py [hunk 1]
```diff
@@ -1,14 +1,63 @@
...
+def workshift_add_multiple_use_case(repo, work_shifts, existing_shifts):
     """
     The function adds multiple work shifts into the chosen database.
+    It checks each shift for overlap and only adds non-overlapping shifts.
     """
+    responses = []
     for work_shift in work_shifts:
-        workshift_add_use_case(repo, work_shift)
+        shift_id = (work_shift.code if isinstance(work_shift, WorkShift)
+            else work_shift["code"])
```
Comment (kate-holdener) [inline server/use_cases/add_workshifts.py:27]: Why are you checking if work_shift is an instance of WorkShift here? This looks like a band-aid, that covers up some internal problem. In this case, the work_shift should always be a JSON object or always be a WorkShift (not sometimes JSON, sometimes WorkShift).

File: server/use_cases/add_workshifts.py [hunk 2]
```diff
@@ -1,14 +1,63 @@
...
+
+    return responses
+
+def shift_already_exists(new_shift, existing_shifts):
+    if isinstance(new_shift, WorkShift):
+        new_shift_start = convert_timestamp_to_datetime(new_shift.start_time)
+        new_shift_end = convert_timestamp_to_datetime(new_shift.end_time)
+    elif isinstance(new_shift, dict):
+        new_shift_start = convert_timestamp_to_datetime(new_shift["start_time"])
+        new_shift_end = convert_timestamp_to_datetime(new_shift["end_time"])
```
Comment (kate-holdener) [inline server/use_cases/add_workshifts.py:44]: Again, this looks like a band-aid fix, to cover up some internal problem. Let's decide of new_shift is always a WorkShift or a dict and remove the calls to isinstance. Generally, if you are finding yourself calling isinstance, it's an indication of some internal problem.

File: server/use_cases/add_workshifts.py [hunk 3]
```diff
@@ -1,14 +1,63 @@
...
+    else:
+        raise ValueError("Invalid shift format")
+
+    for shift in existing_shifts:
+        if isinstance(shift, WorkShift):
+            existing_start = convert_timestamp_to_datetime(shift.start_time)
+            existing_end = convert_timestamp_to_datetime(shift.end_time)
+        elif isinstance(shift, dict):
+            existing_start = convert_timestamp_to_datetime(shift["start_time"])
+            existing_end = convert_timestamp_to_datetime(shift["end_time"])
```
Comment (kate-holdener) [inline server/use_cases/add_workshifts.py:54]: Again, isinstance should not be used here. The shift object should always be either WorkShift or dict.

File: server/use_cases/add_workshifts.py [hunk 4]
```diff
@@ -1,14 +1,63 @@
...
+        else:
+            continue
+        if (max(existing_start, new_shift_start) <
+            min(existing_end, new_shift_end)):
+            return True
+    return False
+
+def convert_timestamp_to_datetime(timestamp_millis):
```
Comment (kate-holdener) [inline server/use_cases/add_workshifts.py:62]: I think this is an unnecessary calculation. You convert the timestamp to datetime, and then compare datetime objects. You can skip the conversion and just compare the timestamps, thus, doing less calculations and making the operation just a little bit faster.

**Review by kate-holdener — CHANGES_REQUESTED** [review]
@Sai9797 I left some comments for you. Please address them. I am especially concerned about the use of isinstance. 
I haven't run this version of the code end-to end, I'll do that after you address the comments in this review.

**Review by kate-holdener — APPROVED** [review]
Looks good!

**Pull request merged**